rust_src/src/minibuf.rs

```diff
@@ -424,4 +427,107 @@ pub fn read_no_blanks_input(
+/// function, instead of the usual behavior.
+///
+/// Optional arg PREDICATE if non-nil is a function limiting the buffers that can
+/// be considered.
+#[lisp_fn(min = "1")]
+pub fn read_buffer(
+    mut prompt: Option<LispStringRef>,
+    mut def: LispObject,
+    require_match: LispObject,
+    predicate: LispObject,
+) -> LispObject {
+    let count = c_specpdl_index();
+
+    if let Some(buffer) = def.as_buffer() {
+        def = buffer.name();
+    }
+
+    unsafe {
+        specbind(
+            Qcompletion_ignore_case,
+            globals.read_buffer_completion_ignore_case.into(),
+        )
+    }
+
+    let result = if unsafe { globals.Vread_buffer_function }.is_nil() {
+        if !def.is_nil() {
+            // A default was provided: PROMT must be changed, editing in the
+            // default value before the colon. To achieve this, PROMT is replaced
+            // with a substring that doesn't contain the terminal space and colon
+            // (if present).
+            if let Some(mut string) = prompt {
+                let data = string.as_slice();
+                let mut len = string.len_bytes() as usize;
+                if len >= 2 && &data[len - 2..len] == &[b':', b' '] {
+                    len -= 2;
+                } else if len >= 1 && (data[len - 1] == b':' || data[len - 1] == b' ') {
+                    len -= 1;
+                }
+                prompt = unsafe {
+                    make_specified_string(
+                        string.sdata_ptr(),
+                        -1,
+                        len as isize,
+                        string.is_multibyte(),
+                    )
+                    .into()
+                };
+            }
+            let format_str = CString::new("%s (default %s): ").unwrap();
+            let format_ptr = format_str.as_ptr();
+            let format =
+                unsafe { make_unibyte_string(format_ptr, format_str.as_bytes().len() as isize) };
+            prompt = editfns::format(&mut [
+                format,
+                prompt.into(),
+                def.as_cons().map_or(def, |d| d.car()),
+            ])
+            .into();
+        }
+        completing_read(
+            prompt.into(),
+            intern("internal-complete-buffer").into(),
+            predicate,
+            require_match,
+            Qnil,
+            Qbuffer_name_history,
+            def,
+            Qnil,
+        )
+    } else if predicate.is_nil() {
+        // Partial backwards compatability for older read_buffer_functions which
+        // expect a 'predicate' argument.
+        call!(
+            unsafe { globals.Vread_buffer_function },
+            prompt.into(),
+            def,
+            require_match
+        )
+    } else {
+        call!(
+            unsafe { globals.Vread_buffer_function },
+            prompt.into(),
+            def,
+            require_match,
+            predicate
+        )
+    };
```

Comment:
I wonder if there is an elegant way to avoid this redundant call of `globals.Vread_buffer_function`.

Reply:
Invoke `funcall` directly. It takes a slice. Something like:
```
else {
    let mut args = vec![
        unsafe { globals.Vread_buffer_function },
        prompt.into(),
        def,
        require_match
    ];
    if predicate.is_not_nil() {
        args.push(predicate)
    }
    eval::funcall(&mut args);
};
```

Reply:
`if predicate.is_not_nil() { predicate } else {}` right ?